ddtrace/internal/utils/formats.py

```diff
@@ -79,21 +80,47 @@ def parse_tags_str(tags_str):
     :param tags_str: A string of the above form to parse tags from.
     :return: A dict containing the tags that were parsed.
     """
-    parsed_tags = {}  # type: Dict[str, str]
     if not tags_str:
-        return parsed_tags
+        return {}
 
     if _TAGS_NOT_SEPARATED.search(tags_str):
         log.error("Malformed tag string with tags not separated by comma or space '%s'.", tags_str)
-        return parsed_tags
+        return {}
 
-    # Identify separator based on which successfully identifies the correct
-    # number of valid tags
-    numtagseps = tags_str.count(":")
-    for sep in [",", " "]:
-        if sum(":" in _ for _ in tags_str.split(sep)) == numtagseps:
-            break
-    else:
+    TAGSEP = ", "
+
+    def parse_tags(tags):
```

Comment:
any reason to define this as a closure?

Reply:
No, except that it is only used as a local helper within this function

Reply:
The function is compiled once when the module is parsed and compiled. Inside the function there is bytecode like so

~~~
  2           0 LOAD_CONST               1 (<code object bar at 0x10183e0e0, file "closure.py", line 2>)
              2 LOAD_CONST               2 ('foo.<locals>.bar')
              4 MAKE_FUNCTION            0
              6 STORE_FAST               0 (bar)
~~~
which creates and stores the closure. We could trade this in for a global lookup, but since this function is only used at startup I think we wouldn't see any big performance improvements. WDYT?

Reply:
Interesting question. Seemingly trustworthy answer from SO:

Each time you actually get around to calling `outer()` -function, that's when a new `inner()` function object is created for that invocation of `outer`.  This binds the already-created `inner` bytecode object to a list of globals, including the value of `tags` as passed into that call to `outer`.
As you can imagine, this is pretty fast, since no parsing is needed, just filling in a quick struct with some pointers to other already-existing objects.

So,
Unless this code is looped **a lot** doesn't sound like it is worth optimizing.

Reply:
If you wanted to write unitests specifically for this method, then it would be worth defining it in the "outer" context. From what I see there are tests that go thru the business logic extensively enough, so don't see this aspect worth a change either.